native/src/main/cpp/sender.cpp

```diff
@@ -0,0 +1,219 @@
+        return;
+    }
+
+    senderClass = env->FindClass(SENDER_CLASS);
+    assert(senderClass != NULL);
+
+    sender = (roc_sender*) get_native_pointer(env, senderClass, thisObj);
+
+    address_unmarshall(env, &address, jaddress);
+
+    if ((port = roc_address_port(&address)) < 0) {
+        jclass exceptionClass = env->FindClass("java/io/IOException");
+        env->ThrowNew(exceptionClass, "Error binding sender");
+        return;
+    }
+
+    if (roc_sender_bind(sender, &address) != 0) {
+        jclass exceptionClass = env->FindClass("java/io/IOException");
+        env->ThrowNew(exceptionClass, "Error binding sender");
+        return;
+    }
+
+    // ephemeral port
+    if (!port) {
+        port = roc_address_port(&address);
+        address_set_port(env, jaddress, port);
+    }
+}
+
+JNIEXPORT void JNICALL Java_com_rocproject_roc_sender_Sender_connect(JNIEnv *env, jobject thisObj, jint portType, jint protocol, jobject jaddress) {
+    roc_address     address;
+    roc_sender*     sender;
+    jclass          senderClass;
+
+    if (jaddress == NULL) {
+        jclass exceptionClass = env->FindClass("java/lang/IllegalArgumentException");
+        env->ThrowNew(exceptionClass, "Bad arguments");
+        return;
+    }
+
+    senderClass = env->FindClass(SENDER_CLASS);
+    assert(senderClass != NULL);
+
+    sender = (roc_sender*) get_native_pointer(env, senderClass, thisObj);
+    address_unmarshall(env, &address, jaddress);
+
+    if (roc_sender_connect(sender, (roc_port_type) portType, (roc_protocol) protocol, &address) != 0) {
+        jclass exceptionClass = env->FindClass("java/io/IOException");
+        env->ThrowNew(exceptionClass, "Error with sender connect");
+    }
+}
+
+JNIEXPORT void JNICALL Java_com_rocproject_roc_sender_Sender_writeFloat(JNIEnv *env, jobject thisObj, jfloat sample) {
+    jclass      senderClass;
+    roc_sender* sender;
+    roc_frame   frame;
+    float       samples[2] = { sample, sample };
+
+    senderClass = env->FindClass(SENDER_CLASS);
+    assert(senderClass != NULL);
+
+    sender = (roc_sender*) get_native_pointer(env, senderClass, thisObj);
+
+    memset(&frame, 0, sizeof(frame));
+
+    frame.samples = samples;
+    frame.samples_size = 2 * sizeof(float);
+
+    if (roc_sender_write(sender, &frame) != 0) {
+        jclass exceptionClass = env->FindClass("java/io/IOException");
+        env->ThrowNew(exceptionClass, "Error sending frame");
+    }
+}
+
+JNIEXPORT void JNICALL Java_com_rocproject_roc_sender_Sender_writeFloats(JNIEnv *env, jobject thisObj, jfloatArray jsamples) {
+    jclass      senderClass;
+    jfloat*     samples;
+    jsize       len;
+    roc_frame   frame;
+
+    if (jsamples == NULL) {
+        jclass exceptionClass = env->FindClass("java/lang/IllegalArgumentException");
+        env->ThrowNew(exceptionClass, "Bad arguments");
+        return;
+    }
+
+    senderClass = env->FindClass(SENDER_CLASS);
+    assert(senderClass != NULL);
```

Comment:
Well, the ```FindClass``` heavyness depends on class hierarchy; since for those roc-java classes the hierarchy is quite basic it should not be so heavy.
BTW your proposed optimization can surely improve performances. I'll add it for all methods in next commit